### PR TITLE
Add generation tests for multimodal generative models

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

It was found that none of the vision LLMs have `GenerationTesterMixin` which is why we had many blind spots, which were not being tested. This PR adds the Mixin to all VLMs and fixes where possible. Some tests are still failing, I am working on it. 

I hope the tests will be much cleaner when we refactor `generate`, because right now I had to add some hacks for tests in models with custom generation.  

EDIT:
Apparently, idefics does not work right now without enabling cache. To not forget in the upcoming rework of `generate`: the problem of idefics is in the special treatment of image attn mask [here](https://github.com/huggingface/transformers/blob/7eb3ba82241c927053689270a0751f4ff5d33c54/src/transformers/models/idefics/modeling_idefics.py#L1565). 

cc @gante

### CODE REVIEW

File: src/transformers/models/idefics/modeling_idefics.py [hunk 1]
```diff
@@ -140,47 +140,45 @@ class IdeficsCausalLMOutputWithPast(ModelOutput):
...
     is_encoder_decoder=False,
     attention_mask=None,
     encoder_outputs=None,
     **model_kwargs,
 ):
-    expanded_return_idx = (
-        torch.arange(input_ids.shape[0]).view(-1, 1).repeat(1, expand_size).view(-1).to(input_ids.device)
-    )
-    input_ids = input_ids.index_select(0, expanded_return_idx)
+    if input_ids is not None:
```
Comment (zucchini-nlp) [inline src/transformers/models/idefics/modeling_idefics.py:150]: For me seems like there is no difference and making it this way enables contrastive decoding in Idefics. Should we ask someone who made idefics to look?

File: src/transformers/models/llava/modeling_llava.py [hunk 2]
```diff
@@ -437,8 +437,6 @@ def forward(
                 inputs_embeds, attention_mask, labels, position_ids = self._merge_input_ids_with_image_features(
                     image_features, inputs_embeds, input_ids, attention_mask, labels
                 )
-                if labels is None:
```
Comment (zucchini-nlp) [inline src/transformers/models/llava/modeling_llava.py:440]: i do not thinks we should be manually adding labels to calculate loss, if None was passed. Removed to pass some of the tests in all Llavas

File: tests/generation/test_utils.py [hunk 3]
```diff
@@ -183,25 +191,25 @@ def _get_encoder_outputs(
...
         output_logits=False,
         output_attentions=False,
         output_hidden_states=False,
         return_dict_in_generate=False,
     ):
+        input_name = model_kwargs.pop("input_name")
+        inputs = model_kwargs.pop(input_name)
         logits_process_kwargs, _ = self._get_logits_processor_and_warper_kwargs(
-            input_ids.shape[-1],
+            model_kwargs["attention_mask"].shape[-1],
```
Comment (zucchini-nlp) [inline tests/generation/test_utils.py:205]: `input ids` can now be any type of "main input", like pixel values. Attention mask may be more reliable

File: tests/generation/test_utils.py [hunk 4]
```diff
@@ -756,38 +777,44 @@ def test_beam_sample_generate_dict_output(self):
...
+                inputs_dict["attention_mask"],
+                model.config,
+                is_vision_model=is_vision_model,
+                num_return_sequences=beam_kwargs["num_beams"],
             )
 
     def test_generate_without_input_ids(self):
-        config, _, _, max_length = self._get_input_ids_and_config()
+        for model_class in self.all_generative_model_classes:
+            config, _, max_length = self._get_input_ids_and_config(model_class)
```
Comment (zucchini-nlp) [inline tests/generation/test_utils.py:791]: added as a loop over all models, so that we can get the `model_class` to infer correct `inputs_dict` format

File: tests/generation/test_utils.py [hunk 5]
```diff
@@ -1289,19 +1354,28 @@ def test_assisted_decoding_sample(self):
...
     def test_generate_with_head_masking(self):
         """Test designed for encoder-decoder models to ensure the attention head masking is used."""
         attention_names = ["encoder_attentions", "decoder_attentions", "cross_attentions"]
         for model_class in self.all_generative_model_classes:
-            config, input_ids, attention_mask, max_length = self._get_input_ids_and_config()
-            # We want to test only encoder-decoder models
-            if not config.is_encoder_decoder:
+            config, inputs_dict, _ = self._get_input_ids_and_config(model_class)
+            # We want to test only encoder-decoder models, also skip enc-dec models which are multimodal
+            if not config.is_encoder_decoder or not hasattr(config, "encoder_layers"):
```
Comment (zucchini-nlp) [inline tests/generation/test_utils.py:1374]: Some multimodal encoder-decoder models (pix2struct) do not have this attribute.

File: tests/test_modeling_common.py [hunk 6]
```diff
@@ -60,6 +60,7 @@
     MODEL_FOR_SEQUENCE_CLASSIFICATION_MAPPING_NAMES,
     MODEL_FOR_TOKEN_CLASSIFICATION_MAPPING_NAMES,
     MODEL_FOR_VIDEO_CLASSIFICATION_MAPPING_NAMES,
+    MODEL_FOR_VISION_2_SEQ_MAPPING_NAMES,
```
Comment (zucchini-nlp) [inline tests/test_modeling_common.py:63]: One question here about mapping. I see that "VISION_2_SEQ_MAPPING" does not contain all of vision-language models that can generate. Is it supposed to be that way or we should update it? Right now it is missing Idefics and Fuyu

**Comment by zucchini-nlp** [comment]
Should be ready for review now. All tests (new added and old ones) are passing for all models. VipLlava will be fixed after rebasing the mentioned PR. 

The biggest problem with multimodal models was the use of different main input names in their custom generate. This behavior should be fixed after the `generate` rework is done.

**Review by gante — COMMENTED** [review]
Stopped reviewing thoroughly in `tests/generation/test_utils.py`. 

I noticed that `model_class` now needs to be passed, and we have a new `prepare_config_and_inputs_for_generation` that redefines a few tested things at a class level. To me, this is a symptom that we're not following the same pattern everywhere, so I have a few questions/comments for you to explore :) 
1. `prepare_config_and_inputs_for_generation` often redefines `inputs_dict["input_name"]`. Why do we need the input name at all? The dictionary from `prepare_config_and_inputs_for_common` should work out of the box in `forward` and `generate` (unless I'm missing something)
2. `prepare_config_and_inputs_for_generation` also pops a few inputs. If the inputs are not used at generation time, `generate` should be able to ignore them, no?
3. As you wrote, `sequence_length = input_ids.shape[-1] // 2` can get us into trouble, let's get rid of it for all models. Let's also use `max_new_tokens` as opposed to `max_length`

[We may need a separate PR to fix these, if the required changes are large]

After these questions are streamlined, I believe you'll need a much smaller diff to enable multimodal generative tests :)

File: src/transformers/models/idefics/modeling_idefics.py (hunk 1 above)
Comment (zucchini-nlp) [reply src/transformers/models/idefics/modeling_idefics.py:150]: yes, the idefics model tests are passing, including slow

File: src/transformers/models/kosmos2/configuration_kosmos2.py [hunk 7]
```diff
@@ -250,7 +250,6 @@ class Kosmos2Config(PretrainedConfig):
         vision_config (`dict`, *optional*):
             Dictionary of configuration options used to initialize [`Kosmos2VisionConfig`].
         latent_query_num (`int`, *optional*, defaults to 64):
-            The number of latent query tokens that represent the image features used in the text decoder component.
```
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/kosmos2/configuration_kosmos2.py:253]: oops, probably removed it by accident

**Comment by zucchini-nlp** [comment]
@gante 

1. Yes, I found that some model's use custom `generate` and the main input name in those cases is either pixel_values. So if we leave it as it is and pass in "input_ids" as first main arg and others as model kwargs, those custom calls complain about incorrect arguments. It would be out of box I believe if we called the main `generate` directly, as I see that we check model forward args and main input names there. However now we cannot get rid of this. Another workaround would be to add separate `_get_inputs_and_config` for each VLMs test, similar to what is happening in the audio models. That way we can hardcode the method to return (`main_input_tensor`, `other_kwargs_in_dict`, `config`, `max_length`) in their specific model testers. But I think it is same level of complication as we have now :(
2. Actually `generate` raises error for unused model kwargs in [this line](https://github.com/huggingface/transformers/blob/fa2c49b00baadd4e7c509c7f241015f6deeb7664/src/transformers/generation/utils.py#L1125). And since some VLMs have required args only for that model, trying to get args from `prepare_config_and_inputs_for_common` by name seemed to be infeasible to generalize, even though it is working for text generative models. 
3. Agreed, we can get rid of it. I just encountered a bunch of new fails on text models, just because of changing seq length so decided to leave it for now. That maybe requires another PR to be merged before this one, to just "remove dividing seq length to two" in text models

**Comment by gante** [comment]
@zucchini-nlp let's open a separate PR that addresses the 3rd point (get rid of the `//2` in `sequence_length = input_ids.shape[-1] // 2`), and chat about VLM tests during our next meeting. I really want to move in the direction of less complexity, even if that implies structural changes in `generate` :)

**Comment by zucchini-nlp** [comment]
@gante Yes, there is a PR already that you approved ahaha (https://github.com/huggingface/transformers/pull/30016)
Just waiting for a reply from @ArthurZucker on some comments to merge
